### DISCUSSION TITLE
[Resolved] Unable to use currying

### DISCUSSION THREAD

**Discussion by QuentinWidlocher** [vlang/v#12704] [Ideas]
<details>
<summary>v doctor output</summary>
<pre>
OS: linux, Ubuntu 20.04.3 LTS
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
CC version: cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

vexe mtime: 2021-11-16 13:48:52
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 045579f

Git version: git version 2.25.1
Git vroot status: weekly.2021.46-8-g045579fd
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 333c14de
</pre>
</details>

**What did you do?**
I tried to mess around with closure and higher order functions to use currying with v.
Here is my script :

```v
struct Person {
	name string
	age int
	sex string
}

fn by_name_eq(n string) fn (Person) bool {
	println('running by_name_eq with name : $n')
	return fn [n] (person Person) bool {
		println('running by_name_eq with name : $n and person : $person')
		return person.name == n
	}
}

named_foo := by_name_eq('Foo')

persons := [
	Person{'Foo', 24, 'M'},
	Person{'Bar', 23, 'M'},
	Person{'FooBar', 23, 'F'},
]

mapped := persons.map(named_foo)

println(mapped)
```

**What did you expect to see?**
```
running by_name_eq with name : Foo
running by_name_eq with name : Foo and person : Person{
    name: 'Foo'
    age: 24
    sex: 'M'
}
running by_name_eq with name : Foo and person : Person{
    name: 'Bar'
    age: 23
    sex: 'M'
}
running by_name_eq with name : Foo and person : Person{
    name: 'FooBar'
    age: 23
    sex: 'F'
}
[true, false, false]
```
 
**What did you see instead?**
```
running by_name_eq with name : Foo
running by_name_eq with name : Foo and person : Person{
    name: 'Foo'
    age: 24
    sex: 'M'
}
running by_name_eq with name : Bar and person : Person{
    name: 'Bar'
    age: 23
    sex: 'M'
}
running by_name_eq with name : FooBar and person : Person{
    name: 'FooBar'
    age: 23
    sex: 'F'
}
[true, true, true]
```

Here, the `n` constant change each time I call `named_foo` instead of always being `Foo`.

Maybe it's not possible to do such a thing in v yet, if so, I'm sorry.

**Comment by te-yd** [comment]
Why are you trying to such a complex thing where it is not supported, I don't recommend using currying in a language where partial function application is not implemented by default, it will cause a huge performance drawback while working on big projects, it also leads to unnecessary creation of multiple function stacks, if you really want to do currying, then ask for the developer to implement it in the core where, 
`fn add(x int, y int) int
{
    return a+b
}`
Should be called both by
`a := add(10)(20)`
And
`a := add(10, 20)`

**Comment by QuentinWidlocher** [comment]
Thanks for your answer, I was just playing around to test the limits of the language.
Closure was an option so currying could have been too ?

I agree that implement it in the core would be the most efficient solution but maybe it's not in the language's scope ?

**Comment by te-yd** [comment]
As for that, you should ask the developer.
  
But adding currying by default will significantly change the entire working of functions, so it's up to the dev and it also needs a lot of work.
      
Please consider removing the bug label, and tag with a "feature request"

**Comment by QuentinWidlocher** [comment]
I'd gladly change the label, but I'm not sure I can

**Comment by te-yd** [comment]
> I'd gladly change the label, but I'm not sure I can

Looks like @JalonSolov did it for ya!
Thanks @JalonSolov

**Comment by QuentinWidlocher** [comment]
With v 0.3.0, the behavior is exactly what I expected, thanks v :heart:
